src/tools/miri/tests/pass-dep/concurrency/linux-futex.rs

```diff
@@ -284,6 +284,79 @@ fn concurrent_wait_wake() {
     assert!(woken > 0 && woken < rounds);
 }
 
+// Crucial test which relies on the SeqCst fences in futex lock/unlock.
+fn concurrent_lock_unlock() {
+    const fn unlocked() -> u32 {
+        0
+    }
+
+    fn locked() -> u32 {
+        unsafe { libc::gettid() }.try_into().unwrap()
+    }
+
+    static FUTEX: AtomicU32 = AtomicU32::new(0);
+    static mut DATA: u32 = 0;
+    static WOKEN: AtomicU32 = AtomicU32::new(0);
+
+    let rounds = 50;
+    for _ in 0..rounds {
+        unsafe { DATA = 0 }; // Reset
+        // Suppose the main thread is holding a lock implemented using futex...
+        FUTEX.store(locked(), Ordering::Relaxed);
+
+        let t = thread::spawn(move || {
+            unsafe {
+                if FUTEX.compare_exchange(unlocked(), locked(), Ordering::Relaxed, Ordering::Relaxed).is_ok() {
+                    // We locked the futex from userspace so make a manual fence.
+                    std::sync::atomic::fence(Ordering::Acquire);
```

Comment:
This still look quite different than the other test?

What the test should do:
- in the main thread, prepare by mark the lock as "I own the lock"
- spawn a thread
- in the spawned thread, call FUTEX_LOCK_PI and check whether we slept or not
- in the main thread, release the lock

Reply:
> in the spawned thread, call FUTEX_LOCK_PI and check whether we slept or not

This is not possible. `FUTEX_LOCK_PI` does not tell us if we slept or not.

I don't think using `FUTEX.compare_exchange()` is an issue. `Mutex` is implemented in this way: first try locking using `compare_exchange` in userspace, and spin for few rounds; if still not unlocked, it will call `FUTEX_LOCK_PI` then. So the logic used here is similar to the logic of `Mutex`.

I think other requirements you mentioned were met in the test code.

Reply:
`compare_exchange` has some very strong ordering effects on the memory model, and the entire point of this test is to exercise the memory model. So it may well be an issue.

> FUTEX_LOCK_PI does not tell us if we slept or not

Isn't that what the `FUTEX_WAITERS` thing tells us?

Reply:
> Isn't that what the `FUTEX_WAITERS` thing tells us?

`FUTEX_WAITERS` is set when there are contentions, i.e. there are waiters waiting on this pi futex. When the child thread acquires the lock, the main thread has already unlocked it, so the bit is cleared.

> compare_exchange has some very strong ordering effects on the memory model, and the entire point of this test is to exercise the memory model. So it may well be an issue.

`Ordering::Releaxed` is used here to avoid the change to semantic.

Reply:
> Ordering::Releaxed, instead of Acquire used in Mutex, is used here to avoid the change to semantic.

Yeah but even a relaxed compare_exchange does non-trivial things wrt ordering.